provider/oci/common/interfaces.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"context"
+
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type ApiClient interface {
+	ociCommon.ConfigurationProvider
+
+	Ping() error
+	GetInstanceVnicAttachments(instanceID instance.Id, compartmentID *string) (ociCore.ListVnicAttachmentsResponse, error)
```

Comment:
It seems a bit muddled that this is a mix of concrete OCI methods, with some methods taking some Juju types, and returning OCI types, and some methods taking Juju types and returning Juju types.

Would it be possible to split this layer wise, so that we would either be presenting a layer that is mostly OCI, vs a layer which is Juju objects?

So things like some of these calls take a context.Context, but other ones just use context.Background internally.

It feels like we could have

OCIAPIClient

which would take context.Context objects, and talk in types of oci*.*Response objects
and then have functionality above that which would hide those details.

Reply:
I will look into splitting this into multiple interfaces. The ```Ping()``` and ```GetInstanceVnicAttachments()``` were written as helper functions, but looking back at this, you are right, it is a bit ugly.

All the OCI methods take in ```context.Context``` which is the interface. We pass in ```context.Background```, which is the implementation of ```context.Context```, because all of the calls we use are async, and short lived. There is little chance any of them will hang in any way, and we need to send a cancel, or set a deadline for any of the requests.